### PR TITLE
feat: major refactor, cleanup and jest tests

### PR DESCRIPTION
**Pull request by pimlie**
I apologize in advance cause it will probably be a pain to review, but this is a major refactor of vue-meta to fully bring it under the nuxt umbrella.After this pr is merged it will use eslint, babel and jest. I optimized most of the code and standardized variable names and file names where possible. Also I added tests so we can now guarantee vue-meta generates exactly the same html on the server side as on the client (well, 
actually its vue-meta's interpretation vs jsdom).

It also includes a workaround to prevent polluting all components with our lifecycle hooks, now it only inserts beforeCreate.

I didnt optimize `getMetaInfo` and `getComponentOption`, I guess you could say that those are two core functons of vue-meta and although I briefly looked into it that really deserves a separate pr. Eg I would really like to remove that deepmerge dependency and I think that should be possible (but not sure yet). Also we shouldnt walk through all components that exists, we can do some smart stuff to only walk branches which contain vue-meta components.

I removed the dependencies for the examples to a separate package.json within that dir and updated the readme, they felt like they were cluttering my efforts trying to get a clean project

The only sensible thing we could/should add are some e2e tests, probably should use nuxt for that somehow as nuxt tests already have functionality for that.

### CODE REVIEW

File: src/client/updaters/attribute.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  const vueMetaAttrString = tag.getAttribute(attribute)
+  const vueMetaAttrs = vueMetaAttrString ? vueMetaAttrString.split(',') : []
+  const toRemove = [].concat(vueMetaAttrs)
+
+  for (const attr in attrs) {
+    if (attrs.hasOwnProperty(attr)) {
+      const val = attrs[attr] || ''
+      tag.setAttribute(attr, val)
+
+      if (vueMetaAttrs.indexOf(attr) === -1) {
```
Comment (manniL) [inline src/client/updaters/attribute.js]:
```suggestion
      if (vueMetaAttrs.includes(attr)) {
```

File: src/client/updaters/tag.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
+// borrow the slice method
+const toArray = Function.prototype.call.bind(Array.prototype.slice)
```
Comment (manniL) [inline src/client/updaters/tag.js]: Would `Array.from` fit here?

File: src/client/updaters/tag.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+  if (tags.length > 1) {
+    // remove duplicates that could have been found by merging tags
+    // which include a mixin with metaInfo and that mixin is used
+    // by multiple components on the same page
+    const found = []
+    tags = tags.filter((x) => {
+      const k = JSON.stringify(x)
+      const res = !found.includes(k)
+      found.push(k)
```
Comment (manniL) [inline src/client/updaters/tag.js]: Do we *have to* push "duplicate tags"?

**Comment by manniL** [comment]
Amazing work :raised_hands:

File: src/client/updaters/tag.js (hunk 2 above)
Comment (pimlie) [reply src/client/updaters/tag.js]: This code runs in the browser and Array.from needs to be polyfilled for older ones, therefor I decided not to touch it. I looked in general at this stuff (eg using babel instead of buble with rollup increases package size already by ~2kb) but not at this detail level.

File: src/client/updaters/tag.js (hunk 3 above)
Comment (pimlie) [reply src/client/updaters/tag.js]: To be honest, I dont know. Or in other words, after reading the description of why this code exists I couldnt be bothered enough (for now) to write a test for it (and there where other areas where more/greater improvements could be made)

File: rollup.config.js [hunk 4]
```diff
@@ -1,27 +1,53 @@
...
-    name: 'VueMeta',
-    banner: `/**
+const banner =  `/**
  * vue-meta v${pkg.version}
  * (c) ${new Date().getFullYear()} Declan de Wet & Sébastien Chopin (@Atinux)
  * @license MIT
  */
 `.replace(/ {4}/gm, '').trim()
+
+const baseConfig = {
```
Comment (pimlie) [inline rollup.config.js]: I have created separate builds for umd and cjs, so I was thinking cant we just strip all ssr stuff from the umd build now? It would make the package quite a bit smaller and you dont need ssr stuff when you use vue-meta from a cdn (but would need to have a look if thats possible at all, maybe our rollup master @pi0 has some suggestions?)

File: src/client/updaters/attribute.js (hunk 1 above)
Comment (pimlie) [reply src/client/updaters/attribute.js]: Yes, changed this one and a couple of others I missed. Thanks

File: src/client/updateClientMetaInfo.js [hunk 5]
```diff
@@ -1,64 +1,75 @@
...
+        getTag(tags, 'body')
+      )
+
+      if (newTags.length) {
+        addedTags[type] = newTags
+        removedTags[type] = oldTags
+      }
+    })
+
+    return { addedTags, removedTags }
```
Comment (pimlie) [inline src/client/updateClientMetaInfo.js]: I think there is still a bug or documentation issue here, the oldTags variable gets added to removedTags but old !== removed. I left some todo comments in the test about this as well, as fixing this could be a breaking change (even though the docs also explictly state removedTags). See eg here: https://github.com/pimlie/vue-meta/blob/feat-jest/test/utils/meta-info-data.js#L74

File: src/client/updaters/tag.js (hunk 2 above)
Comment (pi0) [reply src/client/updaters/tag.js]: Polyfill and browser support is the job of bundler/transpiler, not us! Writing legacy code also adds in the modern bundle too.

File: src/client/refresh.js [hunk 6]
```diff
@@ -12,9 +12,16 @@ export default function _refresh (options = {}) {
...
-    updateClientMetaInfo(options).call(this, info)
-    return info
+  return function refresh() {
+    const metaInfo = getMetaInfo(options, this.$root)
+
+    const tags = updateClientMetaInfo(options, metaInfo)
+
+    // emit "event" with new info
+    if (tags && typeof metaInfo.changed === 'function') {
+      metaInfo.changed.call(this, metaInfo, tags.addedTags, tags.removedTags)
```
Comment (pimlie) [inline src/client/refresh.js]: The docs state that `changed` is called with a `this` context of the component on which the `changed` function is defined. This doesnt seem to be the case, refresh is/was called for the component which triggers the meta update and that component is expicitly set as the `this` context. Because of that, when there are multiple changes in a row the `changed` function will only be called for the last component that called refresh (due to the batchUpdate) and will have the last component that changed as the this context.

Maybe I broke this behaviour with the code refactor by removing all the function wrapping it used, but I dont think so. See my last commit to verify this behaviour: https://github.com/nuxt/vue-meta/pull/319/commits/3cfbf878bbae2e5c2820fd27eec4a872ac241b1d

File: src/client/updaters/tag.js (hunk 2 above)
Comment (pimlie) [reply src/client/updaters/tag.js]: I agree, but in the vue-app-head rfc Atinux had explicit concerns about vue-meta's package size and the polyfill for Array.from is much larger then this slice. And we dont really need all the type checking that Array.from does here, but will change it anyway :)

**Comment by pimlie** [comment]
Some more (after)thoughts and suggested breaking changes

- We should create a single `options` property on `metaInfo` which includes the options like `__dangerouslyDisableSanitizers` and `titleTemplate`. Then we dont have to filter them out every where and can just skip one key.
- Combine `__dangerouslyDisableSanitizers` and `__dangerouslyDisableSanitizersByTagID`
- Rename `__dangerouslyDisableSanitizers` to something more sensible like `raw`
- With the risk of increasing the package size again, I think it could be helpful to add some validation. Currently you can use any tag or attribute combination but there is no check if they are valid. Eg Vue already has utils for this, maybe we could try to re-use those: https://github.com/vuejs/vue/tree/dev/src/platforms/web/util
- Maybe we could use VNode's (or jsdom) for ssr instead of hacking our own html string together? 
- Instead of the changed function we could maybe just $emit an event?

File: src/client/updaters/tag.js (hunk 2 above)
Comment (pimlie) [reply src/client/updaters/tag.js]: Hmm, I checked but buble doesnt add a polyfill for Array.from. I will leave that to you all to decide whether thats ok or not.

Related issue: https://gitlab.com/Rich-Harris/buble/issues/81

File: rollup.config.js [hunk 7]
```diff
@@ -1,27 +1,53 @@
-import commonjs from 'rollup-plugin-commonjs'
+import commonJs from 'rollup-plugin-commonjs'
```
Comment (pi0) [inline rollup.config.js]: Why changing the case of a well-known name?

File: rollup.config.js (hunk 4 above)
Comment (pi0) [reply rollup.config.js]: @pimlie You are right. We can have a different entry point (browser.js) for UMD build. Also adding es build would make it double awesome.

File: src/client/refresh.js (hunk 6 above)
Comment (pi0) [reply src/client/refresh.js]: /cc @Atinux

File: src/client/updateClientMetaInfo.js [hunk 8]
```diff
@@ -1,64 +1,75 @@
...
+
+  const htmlTag = getTag(tags, 'html')
+
+  // if this is not a server render, then update
+  if (htmlTag.getAttribute(ssrAttribute) === null) {
+    // initialize tracked changes
+    const addedTags = {}
+    const removedTags = {}
+
+    Object.keys(newInfo).forEach((type) => {
```
Comment (pi0) [inline src/client/updateClientMetaInfo.js]: If babel is in, can we use `for in` here?

File: src/client/updateClientMetaInfo.js [hunk 9]
```diff
@@ -1,64 +1,75 @@
...
+      // ignore these
+      if ([
```
Comment (pi0) [inline src/client/updateClientMetaInfo.js]: Making this array constant would increase code readability and runtime perf by not creating a new object each time.

File: src/client/updateClientMetaInfo.js [hunk 10]
```diff
@@ -1,64 +1,75 @@
...
+      if (type === 'title') {
+        // update the title
+        updateTitle(newInfo.title)
+        return
+      }
 
-      // emit "event" with new info
-      if (typeof newInfo.changed === 'function') {
-        newInfo.changed.call(this, newInfo, addedTags, removedTags)
+      if (['htmlAttrs', 'headAttrs', 'bodyAttrs'].includes(type)) {
```
Comment (pi0) [inline src/client/updateClientMetaInfo.js]: Same for this.

File: src/client/updaters/attribute.js [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+      const val = attrs[attr] || ''
+      tag.setAttribute(attr, val)
+
+      if (!vueMetaAttrs.includes(attr)) {
+        vueMetaAttrs.push(attr)
+      }
+
+      const keepIndex = toRemove.indexOf(attr)
+      if (keepIndex !== -1) {
+        toRemove.splice(keepIndex, 1)
```
Comment (pi0) [inline src/client/updaters/attribute.js]: Not sure. But can't we a `.filter` in later `toRemove.forEach` ?

File: src/server/generateServerInjector.js [hunk 12]
```diff
@@ -1,25 +1,21 @@
...
+ * @param  {(String|Object|Array<Object>)} data - the data value
+ * @return {Object} - the new injector
+ */
+
+export default function generateServerInjector(options, type, data) {
+  if (type === 'title') {
+    return titleGenerator(options, type, data)
+  }
+
+  if (['htmlAttrs', 'headAttrs', 'bodyAttrs'].includes(type)) {
```
Comment (pi0) [inline src/server/generateServerInjector.js]: Const

File: src/client/updateClientMetaInfo.js (hunk 8 above)
Comment (manniL) [reply src/client/updateClientMetaInfo.js]: @pi0 Whats the benefit of `for in` compared to `.forEach`? :thinking:

**Comment by manniL** [comment]
Regarding CI and Code Coverage: we switched to CircleCI now (thanks to @pi0 :relaxed: )

**Comment by pi0** [comment]
@manniL rebased

**Comment by manniL** [comment]
Comments to the comments:

> We should create a single `options` property on `metaInfo` which includes the options like `__dangerouslyDisableSanitizers` and `titleTemplate`. Then we dont have to filter them out every where and can just skip one key.

Nice idea! But then we would have to deal with overrides and stuff again :thinking: Might be just edge cases though

> Combine `__dangerouslyDisableSanitizers` and `__dangerouslyDisableSanitizersByTagID`

Yup! :+1: 

> Rename `__dangerouslyDisableSanitizers` to something more sensible like `raw`

I think keeping `dangerously` in the name would be preferable as many ppl wouldn't know it opens up attack vectors when used wrong otherwise. Similar to Reacts `dangerouslySetInnerHTML`.

> With the risk of increasing the package size again, I think it could be helpful to add some validation. Currently you can use any tag or attribute combination but there is no check if they are valid. Eg Vue already has utils for this, maybe we could try to re-use those: [vuejs/vue:src/platforms/web/util@`dev`](https://github.com/vuejs/vue/tree/dev/src/platforms/web/util)

That'd be neat! Maybe we can strip them out in a "prod build" somehow so it's only for `dev`? :thinking: 

> Maybe we could use VNode's (or jsdom) for ssr instead of hacking our own html string together?

I'd love to see that, yes!

> Instead of the changed function we could maybe just $emit an event?

Would align more with the "Vue way" of doing things, true :relaxed:

File: src/client/updateClientMetaInfo.js (hunk 9 above)
Comment (pimlie) [reply src/client/updateClientMetaInfo.js]: Regarding runtime perf, you are probably gonna hate me for this but please check these tests: https://jsperf.com/const-vs-inline-array

Although there is a difference its almost negligible (~1%). Especially if you look at the absolute performance of the operation, 830 mio / sec is quite fast (on my machine with Chrome 69 its 830 mio/s for inline and 875 mio/s for const).

For php I know for sure that it actually doesnt matter what you do as the php engine is smart enough to recognize such variables as constants and will treat them as constants for you. It seems v8 is also that kind of smart.

File: rollup.config.js (hunk 7 above)
Comment (pimlie) [reply rollup.config.js]: No reason, except probably just being stuck on some thought process and mindlessly looking at other code will in the back of your head thinking about your problem and then changing this because commonJs would be the proper camelCase version of CommonJS even though it doesnt matter at all but hey I could probably do this to fix that other problem so ctrl+s and close. And then after what feels like 5mins you notice its dark outside and you start feeling to get hungry so you look for some food but you remember you need to go to the shops first because you have no food left but oh fuck its past 20.00 already and the shops are now closed...

File: src/client/updaters/attribute.js (hunk 11 above)
Comment (pimlie) [reply src/client/updaters/attribute.js]: refactored to filter + reduce, makes the code indeed much easier to understand

File: src/server/generateServerInjector.js (hunk 12 above)
Comment (pimlie) [reply src/server/generateServerInjector.js]: fyi, there are ~10 cases in Nuxt where also this inline structure is used (did a quick grep on `].includes`)

File: src/client/updateClientMetaInfo.js (hunk 8 above)
Comment (pimlie) [reply src/client/updateClientMetaInfo.js]: I understood that `for in` is notorious for having a very big polyfill

**Comment by pimlie** [comment]
Comments to some of the comments of the comments:

> Nice idea! But then we would have to deal with overrides and stuff again  Might be just edge cases though

Not sure how that would be different from now, its just adding one extra level? Also with the current deepmerge dependency that shouldnt be really an issue I guess?

> I think keeping `dangerously` in the name would be preferable as many ppl wouldn't know it opens up attack vectors when used wrong otherwise.

The name suggestion `raw` stems from: https://vuejs.org/v2/guide/syntax.html#Raw-HTML So people using Vue should already be familiar with the meaning of raw, if not there is as big of chance they introduce XSS with vue-meta as with vue itself.

I just dont like when a library wants to constantly remind me that something might be dangerous when I've read the docs, know what I am doing and the content I insert is trusted so there is nothing dangerous at all. Eg I'd much prefer to print a warning once in dev mode about this (when its used) then having dangerous used in the var name.

File: src/client/updateClientMetaInfo.js (hunk 8 above)
Comment (pi0) [reply src/client/updateClientMetaInfo.js]: @manniL Code Readability + More ES6 + less mem usage when we use untranspiled version. `Object.keys` create a new Array from keys.

Comment (pi0) [reply src/client/updateClientMetaInfo.js]: PS: We use it in other areas like this:

https://github.com/nuxt/vue-meta/pull/319/files#diff-cba0880c28dd98835c4c104cb1467d57R13

File: src/client/updateClientMetaInfo.js [hunk 13]
```diff
@@ -28,22 +28,22 @@ export default function updateClientMetaInfo(options = {}, newInfo) {
     const addedTags = {}
     const removedTags = {}
 
-    Object.keys(newInfo).forEach((type) => {
+    for (const type in newInfo) {
       // ignore these
       if (metaInfoOptionKeys.includes(type)) {
-        return
+        continue
```
Comment (pimlie) [inline src/client/updateClientMetaInfo.js]: What is your preference with regards to the use of `continue`? Makes it the code better readable or should we just use if..elses?

File: src/client/updateClientMetaInfo.js (hunk 8 above)
Comment (manniL) [reply src/client/updateClientMetaInfo.js]: `for in` is in the standard for long. Do you mean `for of` (introduced with ES6 or so)? 🤔

File: src/server/generators/attribute.js [hunk 14]
```diff
@@ -1,3 +1,5 @@
+import { isUndefined } from '../../shared/typeof'
```
Comment (manniL) [inline src/server/generators/attribute.js]: Could we introduce an alias (`~` or so) so we can get rid of relative paths?

Can do that in a separate PR as well

File: src/client/updateClientMetaInfo.js (hunk 8 above)
Comment (pimlie) [reply src/client/updateClientMetaInfo.js]: Ah yes, I might have. The object.keys have been factored out

File: src/client/updateClientMetaInfo.js (hunk 13 above)
Comment (manniL) [reply src/client/updateClientMetaInfo.js]: I'm a huge fan of early returns/continue ☺️

File: rollup.config.js (hunk 4 above)
Comment (pimlie) [reply rollup.config.js]: This has been added. Not sure if its useful, but to be sure I have added a stub for the inject method so the API is at least complete. Do you prefer I also print a warning or throw an error?

File: src/server/generators/attribute.js (hunk 14 above)
Comment (pimlie) [reply src/server/generators/attribute.js]: Please do as I am not sure how this works actually. Is it a babel feature? Or is it that symlink trick?

Comment (manniL) [reply src/server/generators/attribute.js]: It's a webpack feature ☺️

**Comment by manniL** [comment]
WDYT about a "next" branch so we can work on further improvements split up in several small PRs? ☺️

**Comment by pimlie** [comment]
Fine with me. I also would be interested in what you think the next version of VueMeta should be. Will we release a new major and include breaking changes or not? There are some things I'd like to break, besides things mentioned above there is still some additional function wrapping for `inject` and `refresh` which would be a breaking change if we remove it

**Comment by manniL** [comment]
@pimlie Better discuss that in a new issue ☺️

File: src/client/updateClientMetaInfo.js (hunk 13 above)
Comment (pimlie) [reply src/client/updateClientMetaInfo.js]: Yeah me too, although I find the term `continue` sometimes confusing because you could interpret it as continue the code execution (like when debugging) instead of continue with the next item

**Comment by pimlie** [comment]
I wont add any changes to this pr anymore, I think its best to merge this into next (after a final review) and create separate issues/prs for the removedTags and changed fn issues and/or other new features and breaking changes.

File: src/client/refresh.js (hunk 6 above)
Comment (Atinux) [reply src/client/refresh.js]: Well, I never used this function actually, but I believe this PR will be for next major version.
